### PR TITLE
Define PSA_WANT_xxx_KEY_PAIR_yyy step 2/ECC

### LINKED ISSUE #7771: Define `PSA_WANT_xxx_KEY_PAIR_yyy` step 2/ECC
Following #7614 which introduced new `PSA_WANT_xxx_KEY_PAIR_yyy`, we want to remove uses of the internal `MBEDTLS_PSA_zzz_KEY_TYPE_ECC_KEY_PAIR_LEGACY` (where `zzz` is one or `WANT`, `BUILTIN`) symbol it introduced for ECC.

- [ ] Locate the code related to ECC keypair derivation in `psa_crypto.c` and/or `psa_crypto_ecp.c` and replace the `LEGACY` guard with the appropriate `_DERIVE` macro.
- [ ] Locate tests related to ECC keypair derivation and replace the `LEGACY` guard with the appropriate `_DERIVE` macro.
- [ ] Locate the code related to ECC keypair (random) generation in `psa_crypto.c` and/or `psa_crypto_ecp.c` and  replace the `LEGACY` guard with the appropriate `_GENERATE` macro.
- [ ] Locate tests related to ECC keypair derivation and replace the `LEGACY` guard with the appropriate `_GENERATE` macro.
- [ ] Create new internal macros `MBEDTLS_PSA_zzz_KEY_TYPE_ECC_KEY_PAIR_BASIC` (where `zzz` is one or `WANT`, `BUILTIN`), auto-enabled as soon as the corresponding `USE` macro is (which itself is auto-enabled as soon as anything related to the key type is), and replace all remaining uses of `LEGACY` macros with these macros. (See note below.)
- [ ] Make sure there are components in `all.sh` using `MBEDTLS_PSA_CRYPTO_CONFIG`, one enabling only `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_GENERATE` another only `_DERIVE`, a 3rd one enabling only `USE`, and that they pass `make test`.

Note: the internal `_BASIC` macro reflect that fact that in the short term we're not making it a goal to enable builds without support for import-export, as this is heavily used in tests. We are allowed to include more in the build than what's explicitly requested. In the short term, we want to include support for generate and derive only when explicitly requested, but we'll include support for import-export any time anything else is requested (or equivalently, any time `USE` is requested).

Edit: in the meantime we've decided to rename `PSA_WANT_xxx_KEY_PAIR_USE` to `PSA_WANT_xxx_KEY_PAIR_BASIC`, so I think `MBEDTLS_PSA_WANT_xxx_KEY_PAIR_BASIC` is no longer a great name for the internal macro, BASIC wouldn't mean the same thing for both. At the risk of ending up with long name, I suggest being fully explicit and calling the internal macro `MBEDTLS_PSA_WANT_xxx_KEY_PAIR_BASIC_IMPORT_EXPORT`: the user requested basic, and we translated that to basic+import+export internally since supporting only basic without import-export is not a goal for use right now.

**Depends on:** #7614

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR is the 1st one replacing temporary `LEGACY` symbols with proper `BASIC, IMPORT, EXPORT, GENERATE, DERIVE` ones for EC keys.

Resolves #7771

## PR checklist

- [x] **changelog** not required: already introduced in #7614 - can be extended once #7772 and #7773 are done.
- [x] **backport** not required
- [x] **tests** provided

### CODE REVIEW

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:2656]: Naming: this function doesn't just set a configuration it runs the tests. We should keep the `config_` namespace for helpers that just set the config.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:2672]: I'm sorry but I don't understand what's the purpose of this. We're not trying to test acceleration or drivers here, just a basic build with some features disabled.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:2664]: In terms of what's important, we really need to have a test with RSA generation turned off, because that's a big driver of code size and it's not needed on devices that only import private keys. DH and ECC generation requires only a small amount of code and is usually needed (though not if you're only ever doing static DH), so it's a lot less important if turning those off doesn't work.

But from a systematic testing point of view it does make sense to disable generation for everything. So given a choice of: (disable all generation and not test TLS1.3) vs (disable only RSA generation and test TLS 1.3), I think it's a wash. But disabling USE_PSA_CRYPTO is less clear: users who use the fancy PSA stuff are likely to turn that on.

File: library/psa_crypto.c
Comment (gilles-peskine-arm) [reply library/psa_crypto.c:5731]: Derivation involves private keys and cannot be done with only public key support. So I agree that checking `PUBLIC_KEY` here didn't make sense.

I don't understand why `|| BUILTIN` either. What could make sense is **`&& BUILTIN`** (or just checking `BUILTIN` and not `WANT`) because we haven't implemnted the driver interface for key derivation yet. I don't know if that's necessary: maybe this bit can work independently of the rest of the driver interface. But if it doesn't then checking `BUILTIN` would make sense as a temporary limitation.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I have only reviewed the change to the Python script. I don't currently intend to review the rest.

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:2664]: Note: as it currently stands, the function does this for an individual key type, not all 3 of them at once. From a testing perspective, I think that's good: for example if ECDH key generation is disabled but FFDH generate is enabled, we won't be able to do an ECDHE-ECDSA key exchange with TLS 1.2 with `USE_PSA` but we should be able to do a DHE-RSA key exchange. Similarly, with TLS 1.3, if only one of ECDH or FFDH generate is available, we should still be able to do an ephemeral key exchange and automatically negociate the kind of group that we can actually use. (However, that last part is not true yet, as (1) FFDH support in TLS 1.3 is not merged yet, and (2) it doesn't have proper check for `GENERATE`, which we'll probably want to add later, perhaps as part of #7612 where we clean up other dependencies related to ECDH in TLS.)

Now that I've said that, I think I definitely want to test with `USE_PSA` and TLS 1.3 at some point. Probably not right now, as I guess we'll need to fix a few guards in TLS first (again, can be part of #7612), but in the end.

File: tests/scripts/generate_psa_tests.py
Comment (mpg) [reply tests/scripts/generate_psa_tests.py:324]: Note: this PR only does ECC. (Splitting between ECC, RSA and DH was my attempt at having smaller issues/PRs.)

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:2664]: TBH when I was working for introducing this test I found that there was already a component, `component_build_psa_accel_key_type_ecc_key_pair`, which seem to do what I was looking for so I basically started/copied from there.
That's was mostly an initial attempt, so I'll try to address your comments ;)

File: library/psa_crypto.c
Comment (valeriosetti) [reply library/psa_crypto.c:5731]: Wdyt if I set the following guards:

```
defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE) ||
defined(MBEDTLS_BUILTIN_KEY_TYPE_ECC_KEY_PAIR_DERIVE)
```

The `PSA_WANT` part should be "temporary" until key derivation is added to driver. Does it makes sense to you?

File: tests/suites/test_suite_psa_crypto_driver_wrappers.function
Comment (valeriosetti) [reply tests/suites/test_suite_psa_crypto_driver_wrappers.function:13]: Sorry, my fault: when checking for tests in which I had to replace the LEGACY symbol with the DERIVE one, I searched over test files for "derive" and turned out that `ecjpake_rounds()` uses `psa_key_derivation_setup()`. However I didn't check/thought which kind of key derivation was it supposed to do... Sorry for that

**Comment by gilles-peskine-arm** [comment]
Note the CI is still unhappy.

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_psa_tests.py]: Yes, but at this point `PSA_KEY_TYPE_RSA_KEY_PAIR_GENERATE` should be synonymous with `MBEDTLS_GENPRIME` when drivers aren't involved, so I think we don't need special handling for RSA key generation anymore.

**Review by gilles-peskine-arm — COMMENTED** [review]
I have reviewed the changes in `generate_psa_tests.py`, and they seem sensible to me, but incomplete. The generated files still have many uses of `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR` and `PSA_WANT_KEY_TYPE_RSA_KEY_PAIR` (no DH because the generation scripts [are still missing support](https://github.com/Mbed-TLS/mbedtls/issues/7812)). My understanding is that the goal of this issue is to at least get rid of the ECC ones (but in generation code, it's easier to handle RSA and ECC similarly).

So I'm approving the changes in `generate_psa_tests.py`. But I'm not making a judgement on completeness, and I don't currently intend to review the rest of the PR (but I might do it if no one's available).

File: library/psa_crypto.c
Comment (mpg) [reply library/psa_crypto.c:5731]: Makes sense to me, but I think there should a comment explaining that the `WANT` part is temporary. Or actually, perhaps just use the `WANT` part with a comment that says "this should be BUILTIN really but we don't have driver support for this yet"?

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good, however `git grep KEY_TYPE_ECC_KEY_PAIR_LEGACY` still finds a few things, namely 3 occurrences of `LIBTESTDRIVER1_MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR_LEGACY` in `tests/src/drivers/test_driver_key_management.c`.

Also, I'd like a comment here: https://github.com/Mbed-TLS/mbedtls/pull/7810#discussion_r1239500837 as I think the `#if`s look weird on their own without the context, which we might not be obvious next time someone looks at this.

Finally, the CI has a few complaints of its own.

**Comment by mpg** [comment]
> My understanding is that the goal of this issue is to at least get rid of the ECC ones (but in generation code, it's easier to handle RSA and ECC similarly).

If doing RSA and ECC at once is easier than doing just ECC, because of the script, then this PR should address #7772 too. I'm not sure if that's the case, I'll let you decide, @valeriosetti . (But it looks like some of the CI failures are related to RSA, so perhaps that's an indication.)

**Comment by gilles-peskine-arm** [comment]
> If doing RSA and ECC at once is easier than doing just ECC, because of the script

The script treats key types as uniformly as it can, so treating RSA and ECC differently is likely to require adding temporary `if key_type ...` checks in places where there currently aren't any and there won't be where the work is finished. I don't know if that offsets the savings of only doing ECC elsewhere.

Maybe import/generate/derive would be an easier breakdown than rsa/ecc/dh? Or maybe the automatically generated tests can be updated in one PR, and other tests in other PR?

**Comment by valeriosetti** [comment]
@gilles-peskine-arm @mpg 
In [this commit](https://github.com/Mbed-TLS/mbedtls/pull/7810/commits/f2195317c6a2d1c6b7136a469c24b8178cd0d4d7) I tried to achieve what you asked:

- manage all RSA and ECC dependencies in _all_ generated test files
- handle RSA dependencies with LEGACY symbols and ECC ones with new symbols

Let me know if this is closer to what you were expecting.

**Comment by valeriosetti** [comment]
> The script treats key types as uniformly as it can, so treating RSA and ECC differently is likely to require adding temporary `if key_type ...` checks in places where there currently aren't any and there won't be where the work is finished. I don't know if that offsets the savings of only doing ECC elsewhere.

The fix I introduced in https://github.com/Mbed-TLS/mbedtls/pull/7810/commits/f2195317c6a2d1c6b7136a469c24b8178cd0d4d7 is not as ugly as I initially feared and it can quite easily be removed in #7772

File: include/mbedtls/config_psa.h
Comment (gilles-peskine-arm) [inline include/mbedtls/config_psa.h:1020]: I don't understand why this requires yet another symbol. Why not simply
```
if defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC)
#define PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT
#define PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_EXPORT
#endif
```

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:120]: The code to turn PSA mechanism identifiers (`PSA_ALG_xxx`, `PSA_KEY_TYPE_xxx`, etc.) started out simple, but over the time it's grown more complex: ECC curves, now key pairs split into usage. And in parallel we're likely to want to use that in places other than the systematic operation tests here (e.g. #7803 and follow-ups). There's some similar code bit-rotting in `tests/scripts/set_psa_test_dependencies.py`, which we ran locally to convert some legacy dependencies into PSA dependencies but don't use on a regular basis. We should make a unified interface.

I think this counts as partly on the time budget for driver-only ECC and partly on the time budget for PBKDF2, since those are the two projects that need the feature now. I can help with the Python code.

File: include/mbedtls/config_psa.h
Comment (mpg) [reply include/mbedtls/config_psa.h:1020]: Fair question. The reason I went for yet another symbol is so that if at some point in the future we want to support builds without import or export, the extra symbol should make it clearer to distinguish between "this is really about basic" vs "this has not been reviewed / classified / replaced yet" while we work on that.

If you think it's not worth it, I won't insist on it.

File: tests/scripts/generate_psa_tests.py
Comment (mpg) [reply tests/scripts/generate_psa_tests.py:120]: @gilles-peskine-arm I'm not so familiar with those scripts. Do you think you could define an issue with scope that's sensible and fair to add to one of the driver-only ECC EPICs (and similarly for PBKDF2)? Also, should this wait for #7812 or is it independent enough?

**Review by mpg — COMMENTED** [review]
Note: I've reviewed until 8f83aa1cc4d2677dfe820a26eef086cf8bf5031c and haven't found any additional concerns beyond those raised by the CI.

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [reply tests/scripts/generate_psa_tests.py:120]: I've filed an issue https://github.com/Mbed-TLS/mbedtls/issues/7845 but I have a hard time giving it a proper scope because I don't know what the interface should look like. It's a case where I'll know it if I do it. So more of a time-boxed task, but one that'll require less time if I do it because I wrote most of the existing code.

Comment (gilles-peskine-arm) [reply tests/scripts/generate_psa_tests.py:120]: I don't think it'll make any difference to #7812. It might mean fewer places where #7812 will need to inject special handling, but that special handling would already exist for RSA and ECC so it would be very easy to find.

File: include/mbedtls/config_psa.h
Comment (gilles-peskine-arm) [inline include/mbedtls/config_psa.h:1022]: With this change, when `MBEDTLS_PSA_CRYPTO_CONFIG` is disabled, `MBEDTLS_ECP_C` no longer implies `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE`. As I've mentioned before, I think this is right because the philosophy of `MBEDTLS_PSA_CRYPTO_CONFIG` disabled is that PSA supports the same feature set as legacy, and derivation of ECC keys is not in the legacy feature set.

However, this means that in the default configuration, ECC key pair derivation is no longer available. I think this is an API break.

I see two ways to fix that:

* Make an exception to the general philosophy and have `MBEDTLS_ECP_C` enable `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE`, with a comment explaining why. Advantage: very easy to implement. Downside: makes the configuration automation more complicated to understand. Downside: larger code size for a feature that I think few users need.
* Explicitly enable `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` in the default configuration, with a comment explaining why. This has my preference.

This is also what's causing `test_psa_compliance` to fail on the CI. We can resolve that one by explicitly enabling `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` in `all.sh` (like [in 2.28 we enable CMAC](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-2.28.0/tests/scripts/all.sh#L3040)), if we decide that it isn't enabled in the default configuration anymore.

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:620]: Why do we need this like? Shouldn't `generate_key_dependencies(key.description)` already have `xxx_GENERATE` which implies `xxx_BASIC`?

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:129]: That code is complicated and isn't idiomatic Python. It would be more idiomatic to do the traversal a single time, and replace each element by a list of replacements, and flatten the list. This is one case where the idiomatic Python syntax is weird:
```
new_list = [new_slice
            for dep in dep_list
            for new_slice in tweak_key_pair_dependency(dep) if dep.ends_with('KEY_PAIR') else [dep]]
```
where `tweak_key_pair_dependency` returns a list of dependencies (or more generally an iteratable) (you can handle the else-`[dep]` case inside the auxiliary function). This replaces each element of the list by a list `new_slice` and immediately flattens the nested lists.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I think the Python code gives correct results, but is messier than it needs to be.

File: include/mbedtls/config_psa.h
Comment (valeriosetti) [reply include/mbedtls/config_psa.h:1022]: If I understood correctly both options should be applied in `config_psa.h`, right?
I have some issue understanding the 2nd point though. Are you proposing to always enable `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` when `!MBEDTLS_PSA_CRYPTO_CONFIG` no matter what it is the status of `MBEDTLS_ECP_C`?

File: tests/scripts/generate_psa_tests.py
Comment (valeriosetti) [reply tests/scripts/generate_psa_tests.py:620]: I tried so, but if I modify `DEPENDENCY_FROM_KEY` as follows:
```
DEPENDENCY_FROM_KEY = {
    AES_128BIT_ONLY_DEP_REGEX: AES_128BIT_ONLY_DEP,
    r'ECC_PAIR': ["PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_GENERATE"]
```
then I will have both `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_GENERATE` (expected), but also `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR` which is the legacy one that we don't want.
So I think that if we keep this solution then I'll have to search&remove `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR` from the dependency list. But I cannot get the improvement in this case :/

Comment (gilles-peskine-arm) [reply tests/scripts/generate_psa_tests.py:625]: Yes, but I'd expect `generate_key_dependencies` to do the rewrite, not just add `xxx_GENERATE`. Which means it would need to take `dependencies` as a parameter, but that's ok, this function isn't used anywhere else (yet).

Let's keep this as is for now. I expect I'll do some refactoring soon anyway (because I'm going to need dependency-calculating functions in other work), I can take care of this.

File: tests/scripts/generate_psa_tests.py
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:121]: Again, this is not idiomatic Python: idiomatic Python would be to make this function a generator which calls `yield` on each element.

```
def tweak_key_pair_dependency(dep: str, usage: str) -> Iterator[str]:
    if dep.endswith('RSA_KEY_PAIR'):
        yield re.sub(r'RSA_KEY_PAIR\Z', r'RSA_KEY_PAIR_LEGACY', dep)
    …
```


Let's not drag this on further. I expect I'll refactor the dependency calculation code soon and I can take care of this.

File: include/mbedtls/config_psa.h
Comment (gilles-peskine-arm) [reply include/mbedtls/config_psa.h:1022]: No: we'd make `MBEDTLS_ECP_C` imply `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` in addition to `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_{BASIC,IMPORT,EXPORT}`, with a note that this isn't the usual philosophy of `!MBEDTLS_PSA_CRYPTO_CONFIG` but we're doing it for backward compatibility.

**Review by gilles-peskine-arm — DISMISSED** [review]
I'm approving `generate_psa_tests.py` (except for needing to pass pylint). But there's still the [issue of `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` and backward compatibility](https://github.com/Mbed-TLS/mbedtls/pull/7810/files/f08233e03477d1e233c7bcd79c577f6af1ae0fb3#r1245210746) which is related to the remaining CI failure.

File: include/mbedtls/config_psa.h
Comment (valeriosetti) [reply include/mbedtls/config_psa.h:1022]: Ah ok, so I did not correctly understood the 1st comment. Sorry. I'll do as you suggested.

**Review by mpg — APPROVED** [review]
LGTM assuming the CI is happy too - which it should be thanks to the last commit.

**Review by mpg — APPROVED** [review]
LGTM!

**Comment by valeriosetti** [comment]
I just removed the `need-ci` label because the only error found on the OpenCI is actually due to the CI itself, not to the code. I think this PR is ready for review.

**Review by mprse — COMMENTED** [review]
LGTM.
Just one typo.

**Pull request merged**
